src/client.rs

```diff
@@ -61,10 +64,59 @@ impl Client {
             for e in endpoints.as_ref() {
                 let e = e.as_ref();
                 let channel = if e.starts_with(HTTP_PREFIX) {
+                    #[cfg(feature = "tls")]
+                    if let Some(ref connect_options) = options {
+                        if let Some(_) = connect_options.tls {
+                            return Err(Error::InvalidArgs(String::from(
+                                "TLS options are only supported with HTTPS URLs",
+                            )));
+                        }
+                    }
```

Comment:
I had to think about this for a while: It could be confusing if setting it to the default triggers different behaviour than not setting it. It could be useful in some situations to prevent HTTP connections. But if the endpoints come from configuration then that might or might not be the intended behaviour and the developer of the application might not have thought about it. I would leave it like this but mention it in the documentation of `ConnectOptions::with_tls()`, what do you think?

Reply:
> It could be confusing if setting it to the default triggers different behaviour than not setting it.

Sorry, I didn't catch your meaning. 

 When users invoke `Client::connect()`, there are below (URL & Options) combinations(here take `localhost: 2397` for example):

1. `localhost:2397` & None => no TLS
2. `http://localhost:2397` & None => no TLS
3. `https://localhost:2397` & None => TLS with new TlsOptions
4. `localhost:2397`  & `ConnectOptions::new().with_tls()` => TLS with provided TlsOptions
5. `http://localhost:2397` & `ConnectOptions::new().with_tls()` => Error!
6. `https://localhost:2397` & `ConnectOptions::new().with_tls()` => TLS with provided TlsOptions

Reply:
I was talking about combination 5 with `ClientTlsConfig::new()`, compared to 2. For me it would be more intuitive if the configuration would only be used when needed, and it didn't imply enforcing TLS (which can also be useful in some situations).

It looks like tonic uses a TLS connection if the tls configuration is provided even if the requested scheme is HTTP. That would then lead to an unclear protocol error if the other side does HTTP without TLS, which is to be expected from the URL. I would never have imagined that without taking a closer look at the code.

A clear error message is great, but it's still a runtime thing. Having it in the documentation of `ConnectOptions::with_tls()` would decrease the chances of that happening unintentionally in my opinion. So the behaviour is fine, I would just mention it in the documentation to make it clearer what to expect from it.

Reply:
It's OK to mention it in the document.
By the way, can you test the certificate provided by `ConnectOptions::with_tls()` with `tls` feature (without `tls-roots`)?

Reply:
Just the feature `tls` worked for me with `.with_tls(TlsOptions::new().ca_certificate(Certificate::from_pem(...)))`.

Reply:
Thank you very much!